### DISCUSSION TITLE
Use SSO auth for both directus and the frontend (SDK)?

### DISCUSSION THREAD

**Discussion by ghost** [directus/directus#17408] [Archive]
Let’s say we have a backend (directus.example.com) and a frontend (app.example.com). We are logged in at the backend and now we want to preview an unpublished content (a draft) in the frontend. To make it secure, we need to authenticate at the frontend (via sdk) a second time, right? So we have to put in our login credentials again, even if we are already logged in into directus.

I was wondering if we can make this process easier for clients/users? Like providing a SSO service for directus, so that the user, when logged in into directus, does not have to login on the frontend again — authentication happens in the background.

In other words: Is (seamless) SSO only possible with external services or with directus (user + pass) itself?

**Comment by rijkvanzanten** [comment]
I'm wondering if there's a third option here: Allowing to add the current user's (temporary) access token to the link you're creating, cause that way you could set the button links to something like `https://my-website-preview.com/?token={{ userToken }}` and use that for the preview rendering 🤔 

Alternatively, maybe the cookie set by the Directus app could be configured to also be allowed on your website front-end? That way, you could just refresh the current session to get an access token for the preview usage 🤔

**Reply by ghost** [reply]
> I'm wondering if there's a third option here: Allowing to add the current user's (temporary) access token to the link you're creating, cause that way you could set the button links to something like `https://my-website-preview.com/?token={{ userToken }}` and use that for the preview rendering 🤔

I have also thought of this and actually we could do this already (theoretically) … but I am not sure if that is really save!? If the user (of directus) unintentionally leaks the link including the token, we have a problem. Then I thought, OK, we could hide the token from the addressbar as soon as it arrives at our frontend, but this won’t work if there is an error. So it does not feel right.

---

> Alternatively, maybe the cookie set by the Directus app could be configured to also be allowed on your website front-end? That way, you could just refresh the current session to get an access token for the preview usage 🤔

This would be the best solution for me. Is it possible right now?


---

Another alternative idea is that we use the same domain, but, honestly, I don’t know if this is really possible. The idea is that `app.example.com/admin` is a proxy (rewrite) for `directus.example.com/admin`. So we won’t ever see the URL `directus.example.com`. We could only access it via `app.example.com/admin`. You know what I mean? Then the cookie would have been created under the same domain (`app.example.com` instead of `directus.example.com`) as the app. Does this make sense and is it already possible?

**Reply by rijkvanzanten** [reply]
> I have also thought of this and actually we could do this already (theoretically) … but I am not sure if that is really save!? If the user (of directus) unintentionally leaks the link including the token, we have a problem. Then I thought, OK, we could hide the token from the addressbar as soon as it arrives at our frontend, but this won’t work if there is an error. So it does not feel right.

Yeah, fully agree. The token would be temporary, but it's still not ideal. That being said, we could come up with a way to integrate the same logic that's used for public shares at the moment, which generates out a new temporary token that has locked permissions to only the exact item that was shared (/ previewed) 🤔 

> This would be the best solution for me. Is it possible right now?

I haven't tried it myself yet, but I think it might be! Browsers are pretty finicky when it comes to cookie configuration and sharing though, so it might take some trial and error. I _think_ this might be unlockable by using `SameSite=None` in the cookie, and enforcing it to https only (with `Secure=true`) (both of which can be configured through [the env vars](https://docs.directus.io/self-hosted/config-options.html#security)).

**Reply by ghost** [reply]
> I haven't tried it myself yet, but I think it might be! Browsers are pretty finicky when it comes to cookie configuration and sharing though, so it might take some trial and error. I _think_ this might be unlockable by using `SameSite=None` in the cookie, and enforcing it to https only (with `Secure=true`) (both of which can be configured through [the env vars](https://docs.directus.io/self-hosted/config-options.html#security)).


Is there anyone who already did this or knows how to?

**Reply by claytongulick** [reply]
Our workaround for this was to host the app as a path off the primary domain, instead of on a sub-domain. So if a user logs into ourapp.ourcompany.com/admin the auth works for ourapp.ourcompany.com/someapp too.

We just have nginx deal with the fiddly bits of that.

**Reply by ghost** [reply]
@claytongulick Would you like to share your nginx config?

**Reply by ghost** [reply]
This leads me to an idea: What if we could use the directus app/admin panel as a single page app (spa) inside our frontend. For example, if we could generate an index.html, which we then could copy into our Frontend, so that it is available via frontendapp.com/admin.

This would be great, because auth will happen under the right URL …

@rijkvanzanten What do you think? … and is it already possible (somehow)?

**Reply by claytongulick** [reply]
@JonathanDoelan sure, here it is with domains and stuff renamed. Instead of app1 and app2, just put whatever path you like there. /admin will automatically work with this config.

This is set up to serve the api from the normal root path, and sends all other static requests to the correct app paths. It uses the standard way of setting up PWAs, rewriting all non-file requests to index.html and the app router picks it up from there.

We let certbot manage some lines in here, they're marked.

```
upstream some-server {
	server 127.0.0.1:8055;
	keepalive 32;
}

server {
    server_name some-subdomain.fantastic-company.com; # managed by Certbot

    listen 443 ssl; # managed by Certbot
    ssl_certificate /etc/letsencrypt/live/some-subdomain.fantastic-company.com/fullchain.pem; # managed by Certbot
    ssl_certificate_key /etc/letsencrypt/live/some-subdomain.fantastic-company.com/privkey.pem; # managed by Certbot
    include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot

    #api server and admin
    location / {
        proxy_pass http://some-server;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header REMOTE-HOST $remote_addr;
        add_header X-Cache $upstream_cache_status;

        proxy_connect_timeout 30s;
        proxy_read_timeout 86400s;
        proxy_send_timeout 30s;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
    }

    #first app
    location /app1/ {
	alias /path/to/static/dist/;
	try_files $uri $uri/ /app1/index.html;
    }

    location /app1 {
	alias /path/to/sttatic/dist/;
	try_files $uri $uri/ /app1/index.html;
    }

    #second app
    location /app2/ {
	alias /path/to/static/dist/;
	try_files $uri $uri/ /app2/index.html;
    }

    location /app2 {
	alias /path/to/sttatic/dist/;
	try_files $uri $uri/ /app2/index.html;
    }


}

#redirect port 80 traffic to ssl
server {
    if ($host = some-subdomain.fantastic-company.com) {
        return 301 https://$host$request_uri;
    } # managed by Certbot


	listen 80 ;
    server_name some-subdomain.fantastic-company.com;
    return 404; # managed by Certbot


}
```

**Reply by ghost** [reply]
@claytongulick Thanks!
